### PR TITLE
dlt-connection: add socket timeout

### PR DESCRIPTION
**Pull request by alexmohr**
in some scenarios a socket might block forever.
This means dlt-daemon won't respond properly anymore. To fix this a timeout is added to the dlt-daemon socket connection.

The program was tested solely for our own use cases, which might differ from yours.
Licensed under Mozilla Public License Version 2.0

<sup>Alexander Mohr, <alexander.m.mohr@mercedes-benz.com>, Mercedes-Benz Tech Innovation GmbH, [imprint](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md)</sup>

### CODE REVIEW

**Comment by lvklevankhanh** [comment]
Hi @alexmohr ,
The code change is fine to me. However, could you please share us the way to test socket not responding, and how to verify your code change to solve the problem?
Thanks,

**Comment by alexmohr** [comment]
> Hi @alexmohr , The code change is fine to me. However, could you please share us the way to test socket not responding, and how to verify your code change to solve the problem? Thanks,

I was never able to reproduce the issue in a synthetic test. It only happened from time to time on our hardware target. After changing this the issues are gone but I'm still not able to reproduce it in a unit test

**Comment by lvklevankhanh** [comment]
Thanks @alexmohr for your informative!

File: src/daemon/dlt_daemon_connection.c [hunk 1]
```diff
@@ -402,6 +402,19 @@ int dlt_connection_create(DltDaemonLocal *daemon_local,
         return -1;
     }
 
+    struct timeval timeout;
+    timeout.tv_sec = 5;
+    timeout.tv_usec = 0;
+#ifdef DLT_SYSTEMD_WATCHDOG_ENABLE
+    char *watchdogUSec = getenv("WATCHDOG_USEC");
+    if (watchdogUSec)
+        timeout.tv_sec = atoi(watchdogUSec) / 4000000;
```
Comment (michael-methner) [inline src/daemon/dlt_daemon_connection.c:411]: Shouldn't this be:
```suggestion
    {
        timeout.tv_sec = atoi(watchdogUSec) / 1000000;

        timeout.tv_usec = atoi(watchdogUSec) % 1000000;
   }
```

**Comment by michael-methner** [comment]
Hello @alexmohr ,
i assume that multiple read and write operations to this socket will now cause a timeout as returnvalue / errno . Have you checked that all occurrences handle this in a proper way?

**Comment by alexmohr** [comment]
> Hello @alexmohr , i assume that multiple read and write operations to this socket will now cause a timeout as returnvalue / errno . Have you checked that all occurrences handle this in a proper way?

It's mainly used for the dlt-daemon events which are sent to the clients. 
This PR is related to https://github.com/COVESA/dlt-daemon/pull/385 which I closed because I thought it's super seeded by this PR. It turns out to fully sort out the timeouts both of these patches are necessary. Should I pick the other commit onto this branch so we have one PR to discuss these issues on?

**Comment by michael-methner** [comment]
> > Hello @alexmohr , i assume that multiple read and write operations to this socket will now cause a timeout as returnvalue / errno . Have you checked that all occurrences handle this in a proper way?
> 
> It's mainly used for the dlt-daemon events which are sent to the clients. This PR is related to #385 which I closed because I thought it's super seeded by this PR. It turns out to fully sort out the timeouts both of these patches are necessary. Should I pick the other commit onto this branch so we have one PR to discuss these issues on?

We can keep them as seperate PRs. As they are touching different code lines, it should not be a problem to merge them separately.

File: src/daemon/dlt_daemon_connection.c (hunk 1 above)
Comment (alexmohr) [reply src/daemon/dlt_daemon_connection.c]: Fixed

File: src/daemon/dlt_daemon_connection.c [hunk 2]
```diff
@@ -402,6 +402,21 @@ int dlt_connection_create(DltDaemonLocal *daemon_local,
...
     }
 
+    struct timeval timeout;
+    timeout.tv_sec = 5;
+    timeout.tv_usec = 0;
+#ifdef DLT_SYSTEMD_WATCHDOG_ENABLE
+    char *watchdogUSec = getenv("WATCHDOG_USEC");
+    if (watchdogUSec) {
+        timeout.tv_sec = atoi(watchdogUSec) / 1000000;
+        timeout.tv_usec = atoi(watchdogUSec) % 1000000;
```
Comment (lvklevankhanh) [inline src/daemon/dlt_daemon_connection.c:412]: The fixed is fine.

**Pull request merged**